x/msg_authorization/internal/keeper/keeper.go

```diff
@@ -0,0 +1,130 @@
+package keeper
+
+import (
+	"bytes"
+	"fmt"
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/baseapp"
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/cosmos-sdk/x/msg_authorization/internal/types"
+)
+
+type Keeper struct {
+	storeKey sdk.StoreKey
+	cdc      *codec.Codec
+	router   baseapp.Router
+}
+
+// NewKeeper constructs a message authorisation Keeper
+func NewKeeper(storeKey sdk.StoreKey, cdc *codec.Codec, router baseapp.Router) Keeper {
+	return Keeper{
+		storeKey: storeKey,
+		cdc:      cdc,
+		router:   router,
+	}
+}
+
+func (k Keeper) getActorAuthorizationKey(grantee sdk.AccAddress, granter sdk.AccAddress, msgType string) []byte {
+	return []byte(fmt.Sprintf("c/%x/%x/%s", grantee, granter, msgType))
+}
+
+func (k Keeper) getAuthorizationGrant(ctx sdk.Context, actor []byte) (grant types.AuthorizationGrant, found bool) {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(actor)
+	if bz == nil {
+		return grant, false
+	}
+	k.cdc.MustUnmarshalBinaryBare(bz, &grant)
+	return grant, true
+}
+
+func (k Keeper) update(ctx sdk.Context, grantee sdk.AccAddress, granter sdk.AccAddress, updated types.Authorization) {
+	actor := k.getActorAuthorizationKey(grantee, granter, updated.MsgType())
+	grant, found := k.getAuthorizationGrant(ctx, actor)
+	if !found {
+		return
+	}
+	grant.Authorization = updated
+	store := ctx.KVStore(k.storeKey)
+	store.Set(actor, k.cdc.MustMarshalBinaryBare(grant))
+}
+
+// DispatchActions attempts to execute the provided messages via authorization
+// grants from the message signer to the grantee.
+func (k Keeper) DispatchActions(ctx sdk.Context, grantee sdk.AccAddress, msgs []sdk.Msg) (*sdk.Result, error) {
+	var msgResult *sdk.Result
+	var err error
+	for _, msg := range msgs {
+		signers := msg.GetSigners()
+		if len(signers) != 1 {
+			return nil, sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "authorization can be given to msg with only one signer")
+		}
+		granter := signers[0]
+		if !bytes.Equal(granter, grantee) {
+			authorization, _ := k.GetAuthorization(ctx, grantee, granter, msg.Type())
```

Comment:
I'll need to update the ADR draft but we should use `proto.MessageName(msg)` now instead of `msg.Type()`